docs: add user guide (#1292)
docarray/docarray: docs/user_guide/representing/first_step.md

class BannerDoc(BaseDoc):
    image_url: ImageUrl
    title: str
    description: str
```

You can then instantiate a `BannerDoc` object and access its attributes.

```python
banner = BannerDoc(
    image_url='https://example.com/image.png',
    title='Hello World',
    description='This is a banner',
)

assert banner.image_url == 'https://example.com/image.png'
assert banner.title == 'Hello World'
assert banner.description == 'This is a banner'
```

## `BaseDoc` is a Pydantic `BaseModel`

The class [BaseDoc][docarray.base_doc.doc.BaseDoc] inherits from Pydantic [BaseModel](https://docs.pydantic.dev/usage/models). So you can use
all the features of `BaseModel` in your `Doc` class. 

This means that `BaseDoc`:

* Will perform data validation: `BaseDoc` will check that the data you pass to it is valid. If not, it will raise an 
error. Data being "valid" is actually defined by the type used in the type hint itself, but we will come back to this concept later. (TODO add typing section)
* Can be configured using a nested `Config` class, see Pydantic [documentation](https://docs.pydantic.dev/usage/model_config/) for more detail on what kind of config pydantic offers.

> * Can be used as a drop-in replacement for `BaseModel` in your code and is compatible with tools that use Pydantic like [FastAPI]('https://fastapi.tiangolo.com/').

Reviewer: Just a note to reference FastAPI integration page here that I'm currently working on whoever merges the second

# Representing

At the heart of `DocArray` lies the concept of [`BaseDoc`][docarray.base_doc.doc.BaseDoc].

A [BaseDoc][docarray.base_doc.doc.BaseDoc] is very similar to [Pydantic](https://docs.pydantic.dev/)
[`BaseModel`](https://docs.pydantic.dev/usage/models). It allows to define custom `Document` schema (or `Model` in
the Pydantic world) to represent your data.

## Basic `Doc` usage.

Before going in detail about what we can do with [BaseDoc][docarray.base_doc.doc.BaseDoc] and how to use it, let's

> see how it looks like in practice.

Reviewer 2: ```suggestion
see what it looks like in practice.
```

> The following python code will define a `BannerDoc` class that will be used to represent banner data.

Reviewer 2: ```suggestion
The following Python code defines a `BannerDoc` class that can be used to represent the data of a website banner.
```

* Can be configured using a nested `Config` class, see pydantic [documentation](https://docs.pydantic.dev/usage/model_config/) for more details on what kind of config Pydantic offer.

* Can be used as a drop-in replacement for `BaseModel` in your code and is compatible with tools using Pydantic like [FastAPI]('https://fastapi.tiangolo.com/').

###  What is the difference with Pydantic `BaseModel`? (INCOMPLETE)

LINK TO THE VERSUS (not ready)

[BaseDoc][docarray.base_doc.doc.BaseDoc] is not only a [BaseModel](https://docs.pydantic.dev/usage/models), 

* it allows to be used with DocArray [Typed](docarray.typing) that are oriented toward MultiModal (image, audio, ...) data and for 
Machine Learning use case TODO link the type section. 

Another difference is that [BaseDoc][docarray.base_doc.doc.BaseDoc] has a generated by default `id` field that is used to uniquely identify a document.

## `BaseDoc` allows to represent MultiModal and nested Data.

Let's say you want to represent a Youtube video in your application. Maybe to build a search system for Youtube video.

> A Youtube video is not only composed of a video, but it also has a title, a description, a thumbnail (and more but let's keep it simple).

Reviewer 2: ```suggestion
A Youtube video is not only composed of a video, but it also has a title, a description, a thumbnail (and more, but let's keep it simple).
```

> All of these elements are from different `modalities` LINK TO MODALITIES SECTION (not ready), title and description are text, the thumbnail is an image, and the video in itself is, well, a video.

Reviewer 2: ```suggestion
All of these elements are from different `modalities` LINK TO MODALITIES SECTION (not ready): title and description are text, the thumbnail is an image, and the video in itself is, well, a video.
```

> DocArray allows to represent all of this Multi Modal data in a single object. 

Reviewer 2: ```suggestion
DocArray allows to represent all of this multimodal data in a single object. 
```

Let's first create an `BaseDoc` for each of the elements that compose the Youtube video.

First for the thumbnail which is an image
```python
from docarray import BaseDoc
from docarray.typing import ImageUrl, ImageBytes

class ImageDoc(BaseDoc):
    url: ImageUrl
    bytes: ImageBytes = (
        None  # bytes are not always loaded in memory, so we make it optional
    )
```

Then for the video which is a video
```python
from docarray import BaseDoc
from docarray.typing import VideoUrl, VideoBytes

class ImageDoc(BaseDoc):
    url: VideoUrl

>     bytes: VideoBytes = (

Reviewer 2: same here

Author: No if you put `=None` it is optional by default
